### ISSUE TITLE
Issue with Displaying Chinese Characters

### ISSUE THREAD

**Issue by MacErlang** [SabakiHQ/sgf#7]
Hello,

I am using an iMac, and Sabaki seems to have difficulty displaying Chinese characters properly.
This often occurs with sgf files that I downloaded from the internet. Following comments from
another thread, I have tried to add CA[GB2312] to the file, but it did not work.
A sample file is given below. Can someone enlighten me with a solution to this?

Many thanks in advance,
Shun

~~~
(;AB[pb][pc][pd][pe][qe][rf][sf][qg][pa][qa]AW[ra][rb][qb][qc][qd][re][sd]C[£®“ª£©∆À°¢µπ∆À”Î“™µπ∆À
1£Æ∆À
     ∞◊∆Â‘⁄∫⁄∆Âœ»œ¬µƒ ±∫Ú£¨…˙À¿»Á∫Œƒÿ£ø
]
AP[MultiGo:4.2.1]SZ[19]MULTIGOGM[1]
;B[sb];W[sc];B[rd]C[µƒ∫⁄1µ„£¨∞◊2∂• ±£¨∫⁄3º¥ «∆À£¨];W[rc]C[∞◊4÷ªµ√Ã·£¨∫⁄5≥‘£¨∞◊∆Â±ª…±°£]
;B[se]C[[“™µ„\]£∫À¿ªÓŒ Ã‚÷Æ÷–”√µΩ°∞∆À°±µƒµÿ∑Ω∫‹∂‡£¨∆À « π∂‘∑Ωµƒ—€±‰≥…ºŸ—€µƒ ÷∂Œ°£]
N[“™µ„])
~~~

The actual file is attached below, with added .txt file extension:

[__Vs__9.sgf.txt](https://github.com/SabakiHQ/Sabaki/files/4143251/__Vs__9.sgf.txt)

**Comment by MacErlang** [comment]
Great. Does this mean that the fix will be in the next release?

Thanks,
Shun

**Comment by MacErlang** [comment]
Hello,

Just downloaded and installed the new version, and this problem has not been resolved.
In fact, the new version won't even properly display a file that has been explicitly declared to
be GB2312 encoded. Don't know what is going on? Please help!

I have attached two files, one is original, which won't display properly for either 4.4.3 or 5.0,
and the other has an added GB2312 declaration, which would then load properly for 4.3.3, but
NOT for 5.0.

Thanks,
Shun-Chen

[1020 Test.zip](https://github.com/SabakiHQ/sgf/files/4325245/1020.Test.zip)

**Comment by yishn** [comment]
Weird, the file with the added GB2312 declaration loads fine for me.

**Comment by MacErlang** [comment]
@yishn I have tested several other files with added CA[GB2312], and they all do not
display properly. No idea why my installation is different, as my Mac binary is downloaded
from the release link.

Also, how do I test your new commit with an increased buffer size? Do I need to compile
Sabaki myself? Thanks for your help.

**Comment by yishn** [comment]
Yes, [you'd need to build Sabaki yourself](https://github.com/SabakiHQ/Sabaki/blob/master/docs/guides/building-tests.md).

**Comment by MacErlang** [comment]
@yishn I have compiled Sabaki myself, and the issue persists.
I used the commands:

git clone https://github.com/SabakiHQ/Sabaki
cd Sabaki
npm install
npm run build

The compilation seemed to have worked fine. The executable and a screen dump are here:

https://www.dropbox.com/s/uqtyg9p0uwmos4i/Sabaki%20Compile.zip?dl=0

Thanks for your help,
Shun-Chen

**Comment by MacErlang** [comment]
Hi there,

Shun-Chen Niu (scniu@sbcglobal.net) invited you to view the file " Sabaki Compile.zip " on Dropbox.

View file[1]

Enjoy!
The Dropbox team 
 Shun-Chen and others will be able to see when you view this file. Other files shared with you through Dropbox may also show this info. Learn more[2] in our help center. 

[1]: https://www.dropbox.com/l/scl/AAB2c78RKHqDY_CUsWOj4Fk8aUUJgM0g_QA
[2]: https://www.dropbox.com/l/AADIYPvImROQv58Cm-ifer-7tcC-wy5Gr1w

**Comment by yishn** [comment]
After investigation, it seems we're accidentally excluding the decoding library from our bundle. This should be fixed on Sabaki master now. Can you pull, rebuild, and see if the problem is now fixed?

**Comment by MacErlang** [comment]
Just compiled again, and it now works fine with and WITHOUT the GB2312
declaration! Great detective work and many thanks.

**Comment by MacErlang** [comment]
@yishn Sorry to bother you again, but the new version still seems to have issues.
I am attaching two files, one named Original.sgf and the other GB2312.sgf. The Original
does not have any character declaration, and the other one has. What appears to be an
anomaly is that Original.sgf loads fine into 43.3 but does NOT display properly in 50.1. The
file GB2312.sgf does load fine in both versions of Sabaki.

So, there appears to be a discrepancy between the two Sabaki versions. The attached file
is fairly simple, so this is rather strange. Any ideas?

Best,
Shun

[Sample.zip](https://github.com/SabakiHQ/sgf/files/4333703/Sample.zip)

**Comment by yishn** [comment]
Hmm... it seems like detecting encoding on spliced test buffers didn't really work. Now we're just falling back to detecting encoding on the first 1000 bytes of the buffer.

**Comment by MacErlang** [comment]
Just compiled after the new commit. The file Original.sgf still does not display properly.

**Comment by MacErlang** [comment]
In trying to figure out what might have gone wrong, I have inspected lots of files, using the
newly compiled version. What is really weird is that the detection does not seem to be consistent.
I have attached two files, one labeled as Good and the other as Bad. As far as I could tell, the two
files are essentially identical in form, and yet one displays properly and the other does not.
Don't know if this might help you pin down the issue or not.

Shun

[Samples.zip](https://github.com/SabakiHQ/sgf/files/4334478/Samples.zip)

**Comment by MacErlang** [comment]
Let me add that both files in Samples.zip load properly in 43.3.

**Comment by yishn** [comment]
In v0.43.3, we're guessing encoding based on the first 100 bytes of the files. After extending the encoding guessing to the first 1000 bytes of the file, it doesn't guess GB2312 anymore because, as @fohristiwhirl pointed out, "SGF naturally contains a bunch of UTF-8 looking stuff like B[cc]; W[dd]".

If we restrict ourselves to the first 100 bytes again, your [original file](https://github.com/SabakiHQ/sgf/issues/7#issuecomment-581103143) would have issues, because in there, the first 100 bytes doesn't contain any Chinese.

For short term, we can probably just pick something between 100 bytes and 1000 bytes and guess encoding based on that. For long term, [we should let the user pick their own encoding](https://github.com/SabakiHQ/Sabaki/issues/655).

**Comment by MacErlang** [comment]
Does it make sense to let the detection scheme focus on the C[], PB[], and PW[] fields? These are areas where different encoding might make a difference (especially C[]).

**Comment by yishn** [comment]
Yes, that was what we were doing before using spliced test buffers. But that doesn't work as evidenced by [your previous samples](https://github.com/SabakiHQ/sgf/issues/7#issuecomment-599158088).

**Comment by MacErlang** [comment]
Hi there,

In case you missed it, Shun-Chen Niu (scniu@sbcglobal.net) shared "Sabaki Compile.zip" with you on Dropbox.

View on Dropbox[1]

Thanks!
- The Dropbox Team

____________________________________________________
 Dropbox, Inc. 
 PO Box 77767, San Francisco, CA 94107 
View Privacy Policy[2] | Unsubscribe[3]

[1]: https://www.dropbox.com/l/scl/AAD2z69yY2aSZz_8bRLoXS7DWMG9o4GsLHs
[2]: https://www.dropbox.com/l/AADecDnqKGixOtpidMGJUpdgJ6_FUwayCvE/privacy#privacy
[3]: https://www.dropbox.com/l/AAA1Z1pc9TTCMt0_yh6h7Dd0Is_yn0wRshE

**Comment by MacErlang** [comment]
@yishn I have tested some more files, and I am attaching four of them. These are all original
files I downloaded from the web; the only exception is that I added GB[2312] into the file
"倒垂莲（共二变）- GB2312.sgf". All of these files have trouble with either 50.1 or 43.3, in that
most would hang Sabaki. However, I have found out that they ALL load just fine in Version 35.1.
So, I am wondering if the "older" scheme used in 35.1 might be more robust than what has been
implemented in the recent versions?

BTW, I compiled the latest version, but noticed that the new option on user encoding selection
failed to commit properly.

Best,
Shun

[Test Files.zip](https://github.com/SabakiHQ/sgf/files/4346123/Test.Files.zip)
